networkx/algorithms/community/modularity_max.py

```diff
@@ -234,7 +220,115 @@ def greedy_modularity_communities(G, weight=None, resolution=1, n_communities=1)
+        as a weight.  If None, then each edge has weight 1.
+        The degree is the sum of the edge weights adjacent to the node.
+
+    resolution : float, optional (default=1)
+        If resolution is less than 1, modularity favors larger communities.
+        Greater than 1 favors smaller communities.
+
+    cutoff : int or None, optional (default=None)
+        stop iterating when reaching ``cutoff`` community size: the community merging
+        process is terminated once this number of communities is reached, or until
+        modularity can not be further increased. Must be between 1 and the
+        total number of nodes in `G`. Default is ``None``, meaning the community
+        merging process continues until all nodes are either in (a) the same community
+        or (b) until the best community structure is found.
+
+    best_n_communities : int or None, optional (default=None)
+        the merging proceeds continues until the number of communities
+        ``best_n_communities`` is reached, even if this does not further improve the
+        community structure. In this case those nodes are aggregated that decrease
+        modularity the least.
+
+    Returns
+    -------
+    partition: list
+        A list of frozensets of nodes, one for each community.
+        Sorted by length with largest communities first.
+    """
+    if cutoff is not None:
+        if (cutoff < 1) or (cutoff > G.number_of_nodes()):
+            raise ValueError(
+                f"cutoff must be between 1 and {G.number_of_nodes()}. Got {cutoff}"
+            )
+    if best_n_communities is not None:
+        if (best_n_communities < 1) or (best_n_communities > G.number_of_nodes()):
+            raise ValueError(
+                f"best_n_communities must be between 1 and {G.number_of_nodes()}. "
+                f"Got {best_n_communities}"
+            )
+
+    if "n_communities" in aliases:
+        import warnings
+
+        warnings.warn(
+            "kwarg ``n_communities`` in greedy_modularity_communities is deprecated. "
+            "Use ``cutoff`` instead.",
+            DeprecationWarning,
+        )
+        cutoff = aliases.pop("n_communities")
+
+    # retrieve generator object to construct output
+    community_gen = greedy_modularity_communities_generator(
+        G, weight=weight, resolution=resolution
+    )
+
+    # construct the first two best partitions
+    communities_prv = community_gen.__next__()
+    if len(communities_prv) == 1:
+        return communities_prv
+    communities_nxt = community_gen.__next__()
+
+    # calculate modularity for these partitions/communities
+    modularity_prv = modularity(
+        G, communities_prv, resolution=resolution, weight=weight
+    )
+    modularity_nxt = modularity(
+        G, communities_nxt, resolution=resolution, weight=weight
+    )
+    dq = modularity_nxt - modularity_prv
+
+    while len(communities_prv) > 1:
```

Comment:
It seems like this loop should stop based on `communities_nxt` instead of `communities_prv`.
Aren't you yielding one more set of communities than you need?  
I think the other way you also don't have to construct both `nxt` and `prv` before entering the loop.
(I also shorten names here.  See which you prefer. Also I think the argument name might be better
as `best_n` than `best_n_communities` because we expect this computation to be about the communities.)
It's something like: 
```python
    if cutoff is None:
        cutoff = 1
    if best_n is None: 
        best_n = len(G)
    compute `nxt` and `mod_nxt` and set dq=0 to start??
    while len(nxt) > 1 and not (dq > 0 and len(nxt) <= cutoff) and not (dq < 0 and len(nxt) <= best_n):
        prv, mod_prv = nxt, mod_nxt
        construct nxt and mod_nxt
        dq = mod_nxt - mod_prv
    return sorted(nxt, key=len, reverse=True)
```
Much of this is a matter of taste of course, but if it is more readable, that's important.

Reply:
I think not. Because, if `dq<0`, then we want to return `prv`, not `nxt`. In this case `nxt` has already a worse community structure than `prv`, therefore we want to discard it (unless it is specified in best_n to continue merging). Therefore, `prv` should be defined before the loop.
If you do not want to define it before the loop, what you can do, is define only `nxt` before the loop, but still always return `prv`. However, that case, you'd have to make a distinction after the loop, if `len(prv) = 2` (because here now `len(nxt)=1` and the loop automatically breaks) and then check if `nxt` or `prv` should be returned based on dq.

Not sure if I'm overseeing something

Reply:
adapted the rest, though